Allow recommended badge to be a link on the SearchResult component

bobsilverberg opened the pull request:
Fixes #8102

willdurand commented on src/ui/components/RecommendedBadge/index.js:
```diff
@@ -16,31 +17,43 @@ type Props = {|
 |};
 
-export const RecommendedBadgeBase = ({ i18n, size }: InternalProps) => {
+export const sumoLink = 'https://support.mozilla.org/';
```
Maybe this should go to the `constants.js` file?

willdurand commented on src/ui/components/RecommendedBadge/index.js:
```diff
@@ -16,31 +17,43 @@ type Props = {|
   const label = i18n.gettext('Recommended');
 
+  const openSumoLink = (event: SyntheticEvent<HTMLButtonElement>) => {
+    _window.open(sumoLink, '_blank');
```
I was wondering how one could fix the issue with a `<button />`. I have mixed feelings about this _hack_. Ideally, we should fix the issue so that the links works without JavaScript. I'd suggest that we follow what is described here: https://www.sarasoueidan.com/blog/nested-links/.

That would fix another issue I noticed while testing this patch: the button does not look like a link, it shows the default cursor and not the pointer.

I think that will be a little more work, but that would be more future-proof (I believe). WDYT?

bobsilverberg replied:
That's an interesting article. I had read about the `<object>` hack and didn't like the looks of it. I will try implementing what she suggests in the article and see if I can get it to work in a satisfactory manner.

bobsilverberg replied:
@willdurand I was trying to understand exactly what she is doing in the article, but I don't think I quite do. I'm having trouble getting any of it to work. I tried the second to last example, which seems to be syntactically correct, but doesn't do what is described. Then I looked at the final example, and I cannot make sense of it as it contains invalid style syntax (e.g., `top: ...;`). So I'm kind of stuck. Do you think you could take a stab at making it work, or even provide me with a simple example that _does_ work?

bobsilverberg commented:
> What about changing the approach in `SearchResult`? Some ideas:
> 
>     * Instead of wrapping all content in an anchor, add an `onClick` event in the react component and `:hover { cursor: pointer }` to make it look the same as before
> 

This will make the link require JS, won't it? That seemed to be something @willdurand was suggesting we avoid.

>     * Make the `<h2>` a proper link to the detail page, for SEO
> 

The badge is currently inside the `<h2>`, so we'd still have the link inside a link issue. Or are you suggesting we move it outside of the `<h2>`?
>

willdurand commented:
Since this is a known problem that can be solved without JS, I'd prefer to keep actual links instead of requiring JS 😺

kumar303 commented:
> This will make the link require JS, won't it? That seemed to be something @willdurand was suggesting we avoid.

@bobsilverberg I wasn't suggesting to get rid of the links, just to move them to the `h2` and put a click handler on the rest to simulate the old link behavior. This will keep the page accessible and will preserve SEO, too.

Here's an example to illustrate what I meant:

<details>
<summary>patch</summary>

```diff
diff --git a/src/amo/components/SearchResult/index.js b/src/amo/components/SearchResult/index.js
index dc8f55eba..9769464ec 100644
--- a/src/amo/components/SearchResult/index.js
+++ b/src/amo/components/SearchResult/index.js
@@ -60,6 +60,7 @@ export class SearchResultBase extends React.Component<InternalProps> {
       _config,
       _isAllowedOrigin,
       addon,
+      addonInstallSource,
       clientApp,
       i18n,
       showMetadata,
@@ -73,18 +74,32 @@ export class SearchResultBase extends React.Component<InternalProps> {
     const iconURL = getAddonIconUrl(addon);
 
     let imageURL = iconURL;
+    let addonTitle = <LoadingText />;
 
-    if (addon && isTheme(addon.type)) {
-      let themeURL = getPreviewImage(addon);
-
-      if (!themeURL && addon && addon.type === ADDON_TYPE_THEME) {
-        themeURL =
-          addon.themeData && _isAllowedOrigin(addon.themeData.previewURL)
-            ? addon.themeData.previewURL
-            : null;
+    if (addon) {
+      let linkTo = `/addon/${addon.slug}/`;
+      if (addonInstallSource) {
+        linkTo = addQueryParams(linkTo, { src: addonInstallSource });
       }
 
-      imageURL = themeURL;
+      addonTitle = (
+        <Link to={linkTo} className="SearchResult-link">
+          {addon.name}
+        </Link>
+      );
+
+      if (isTheme(addon.type)) {
+        let themeURL = getPreviewImage(addon);
+
+        if (!themeURL && addon && addon.type === ADDON_TYPE_THEME) {
+          themeURL =
+            addon.themeData && _isAllowedOrigin(addon.themeData.previewURL)
+              ? addon.themeData.previewURL
+              : null;
+        }
+
+        imageURL = themeURL;
+      }
     }
 
     // Sets classes to handle fallback if theme preview is not available.
@@ -138,7 +153,7 @@ export class SearchResultBase extends React.Component<InternalProps> {
 
         <div className="SearchResult-contents">
           <h2 className="SearchResult-name">
-            {addon ? addon.name : <LoadingText />}
+            {addonTitle}
             {showRecommendedBadge &&
             _config.get('enableFeatureRecommendedBadges') &&
             addon &&
@@ -203,7 +218,7 @@ export class SearchResultBase extends React.Component<InternalProps> {
   }
 
   render() {
-    const { addon, addonInstallSource } = this.props;
+    const { addon } = this.props;
 
     const result = this.renderResult();
     const resultClassnames = makeClassName('SearchResult', {
@@ -211,20 +226,12 @@ export class SearchResultBase extends React.Component<InternalProps> {
       'SearchResult--persona': addon && addon.type === ADDON_TYPE_THEME,
     });
 
-    let item = result;
-    if (addon) {
-      let linkTo = `/addon/${addon.slug}/`;
-      if (addonInstallSource) {
-        linkTo = addQueryParams(linkTo, { src: addonInstallSource });
-      }
-      item = (
-        <Link to={linkTo} className="SearchResult-link">
-          {result}
-        </Link>
-      );
-    }
-
-    return <li className={resultClassnames}>{item}</li>;
+    // TODO: Add a click handler to simulate the old link behavior.
+    return (
+      <li onClick={() => {}} className={resultClassnames}>
+        {result}
+      </li>
+    );
   }
 }
 
```

</details>

bobsilverberg commented:
Thanks @kumar303, that was really helpful. I have implemented the changes as you suggested and added an `onClick` handler to the `<li>`.  I had an eslint issue where it didn't like me putting an `onClick` on the `<li>`, which I got around by adding some comments, but I'm not sure if there's something else we should do about that.

bobsilverberg commented:
Hmm, I just noticed one unfortunate side effect while testing this. When you click the "Recommended" badge it also triggers the `onClick` of the `<li>` so it still navigates you to the detail page. I'm not sure that that's what we want, but I'm also not sure how to prevent it.

bobsilverberg commented:
> Hmm, I just noticed one unfortunate side effect while testing this. When you click the "Recommended" badge it also triggers the `onClick` of the `<li>` so it still navigates you to the detail page. I'm not sure that that's what we want, but I'm also not sure how to prevent it.

I think I figured out how to deal with this. I'll push a new commit after I have written a test for it.

kumar303 commented on src/amo/components/SearchResult/index.js:
```diff
@@ -202,39 +230,44 @@ export class SearchResultBase extends React.Component<InternalProps> {
-    return <li className={resultClassnames}>{item}</li>;
+    return (
+      // eslint-disable-next-line max-len
+      // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions, jsx-a11y/click-events-have-key-events
```
These are helpful lint errors but I think it's fine to disable them. You could add a comment saying that the link is still keyboard accessible because an actual anchor tag was added to the `h2`. I don't think it's worth adding `onKeyUp` like one of the lint rules is asking you to.

kumar303 commented on src/amo/components/SearchResult/styles.scss:
```diff
@@ -7,6 +7,10 @@ $icon-default-size: 32px;
   list-style-type: none;
   margin: 0 0 1px;
+
+  :hover {
```
The whole element is clickable so it should be:

```diff
diff --git a/src/amo/components/SearchResult/styles.scss b/src/amo/components/SearchResult/styles.scss
index 1f5b3813a..27f301bd8 100644
--- a/src/amo/components/SearchResult/styles.scss
+++ b/src/amo/components/SearchResult/styles.scss
@@ -8,7 +8,7 @@ $icon-default-size: 32px;
   list-style-type: none;
   margin: 0 0 1px;
 
-  :hover {
+  &:hover {
     cursor: pointer;
   }
 }
```

kumar303 commented on src/amo/components/SearchResult/index.js:
```diff
@@ -73,6 +89,18 @@ export class SearchResultBase extends React.Component<InternalProps> {
+
+    if (addon) {
+      addonTitle = (
+        <Link
```
This also needs `stopPropagation()` otherwise the link loads twice and that means you have to press the back button twice to go back.

```diff
diff --git a/src/amo/components/SearchResult/index.js b/src/amo/components/SearchResult/index.js
index 1f1ad5526..bb45d8011 100644
--- a/src/amo/components/SearchResult/index.js
+++ b/src/amo/components/SearchResult/index.js
@@ -96,6 +96,7 @@ export class SearchResultBase extends React.Component<InternalProps> {
         <Link
           to={this.getAddonLink(addon, addonInstallSource)}
           className="SearchResult-link"
+          onClick={(e) => e.stopPropagation()}
         >
           {addon.name}
         </Link>
```

bobsilverberg replied:
Ah, I thought I noticed something odd with the back button while I was testing this, but I wasn't quite sure what it was. Thanks!

kumar303 commented on src/amo/components/SearchResult/styles.scss:
```diff
@@ -7,6 +7,10 @@ $icon-default-size: 32px;
+  }
 }
 
 .SearchResult-link {
```
This needs some tweaks to get the h2 link looking how it did before (when everything was wrapped in a link).

```diff
diff --git a/src/amo/components/SearchResult/styles.scss b/src/amo/components/SearchResult/styles.scss
index 1f5b3813a..36c3d20c8 100644
--- a/src/amo/components/SearchResult/styles.scss
+++ b/src/amo/components/SearchResult/styles.scss
@@ -14,8 +14,19 @@ $icon-default-size: 32px;
 }
 
 .SearchResult-link {
-  color: $text-color-default;
   text-decoration: none;
+
+  &,
+  &:active,
+  &:link,
+  &:visited {
+    color: $text-color-default;
+  }
+
+  &:focus,
+  &:hover {
+    color: $link-color;
+  }
 }
 
 .SearchResult-result {
@@ -128,11 +139,6 @@ $icon-default-size: 32px;
     }
   }
 
-  .SearchResult-link:focus &,
-  .SearchResult-link:hover & {
-    color: $link-color;
-  }
-
   .LoadingText {
     margin-bottom: 5px;
   }
```

bobsilverberg replied:
Thanks! I also had to add:

```diff

diff --git a/src/amo/components/SearchResult/styles.scss b/src/amo/components/SearchResult/styles.scss
index 2f455eeb5..59ad7d729 100644
--- a/src/amo/components/SearchResult/styles.scss
+++ b/src/amo/components/SearchResult/styles.scss
@@ -10,6 +10,10 @@ $icon-default-size: 32px;
 
   &:hover {
     cursor: pointer;
+
+    .SearchResult-link {
+      color: $link-color;
+    }
   }
 }
```

In order to make the link change colour when the user hovers anywhere over the `li`, which is how it behaves in the current version.

kumar303 commented on src/ui/components/RecommendedBadge/index.js:
```diff
@@ -22,6 +22,10 @@ type InternalProps = {|
 export const RecommendedBadgeBase = ({ i18n, size }: InternalProps) => {
   const label = i18n.gettext('Recommended');
 
+  const onClick = (event: SyntheticEvent<any>) => {
```
This is risky because you might not want to stop propagation in all situations. I suggest making it a prop so you can just add the `onClick` handler for `SearchResult` only.

Example:

```diff
diff --git a/src/amo/components/SearchResult/index.js b/src/amo/components/SearchResult/index.js
index 1f1ad5526..049cfc6b2 100644
--- a/src/amo/components/SearchResult/index.js
+++ b/src/amo/components/SearchResult/index.js
@@ -172,7 +172,10 @@ export class SearchResultBase extends React.Component<InternalProps> {
             addon &&
             addon.is_recommended &&
             clientApp !== CLIENT_APP_ANDROID ? (
-              <RecommendedBadge size="small" />
+              <RecommendedBadge
+                onClick={(e) => e.stopPropagation()}
+                size="small"
+              />
             ) : null}
           </h2>
           {summary}
```

kumar303 commented on tests/unit/ui/components/TestRecommendedBadge.js:
```diff
@@ -26,4 +30,12 @@ describe(__filename, () => {
     expect(root).toHaveClassName(className);
   });
+
+  it('Calls event.stopPropagation after clicking on the link', () => {
```
Since specs start with `it` when you read them, we typically don't capitalize the first word.

bobsilverberg commented:
Thanks for the review @kumar303, this is ready for another look.
